### PR TITLE
Support constraint without version in GMM

### PR DESCRIPTION
**Pull request by ljacomet**

### CODE REVIEW

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java [hunk 1]
```diff
@@ -168,7 +168,7 @@ public DependencyConstraint create(Object dependencyNotation, Action<? super Dep
 
     @Override
     public DependencyConstraint enforcedPlatform(Object notation) {
-        DependencyConstraintInternal platformDependency = (DependencyConstraintInternal) create(notation);
+        DependencyConstraintInternal platformDependency = (DependencyConstraintInternal) doCreate(notation, null);
```
Comment (octylFractal) [inline subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java:171]: This strikes me as odd, there isn't much saved in doing this, and it adds a second place that the default `null` is written out.

**Review by octylFractal — APPROVED** [review]
LGTM, just one minor change I had a question about but it's not blocking.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java:171]: I was wondering about it when changing it. But it seemed to make sense to go directly for internal APIs from public ones. But it is true that this instance does not resolve anything.

**Review by hegyibalint — APPROVED** [review]

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradleModuleMetadataParser.java [hunk 2]
```diff
@@ -439,7 +439,7 @@ private IvyArtifactName consumeArtifactSelector(JsonReader reader) throws IOExce
             String group = null;
             String module = null;
             String reason = null;
-            VersionConstraint version = null;
+            VersionConstraint version = DefaultImmutableVersionConstraint.of();
```
[the remark this answers is not in the document]
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/ivyservice/ivyresolve/parser/GradleModuleMetadataParser.java:442]: Indeed, but the assumption made here is wrong. The Gradle DSL allows to define a constraint without version. The whole dependency graph engine works fine with a constraint that has an _empty_ version. And there are limited use cases for this - see the added test.

What is clearly not supported by the engine is a `null` version, this is fixed by this change.

File: subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java (hunk 1 above)
Comment (ljacomet) [reply subprojects/dependency-management/src/main/java/org/gradle/api/internal/artifacts/dsl/dependencies/DefaultDependencyConstraintHandler.java]: I ended up reverting that change

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Comment by blindpirate** [comment]
So the ~10% performance regression seems to be true...

**Comment by blindpirate** [comment]
@bot-gradle test and merge

**Pull request merged**